national_rail/transform.py

```diff
@@ -0,0 +1,163 @@
+""" Module for transforming data received from the National Rail API into a 
+    useful format. """
+
+import logging
+
+from datetime import datetime, timedelta
+import xml.etree.ElementTree as ET
+from bs4 import BeautifulSoup
+
+
+def read_data_from_file(filename: str) -> str:
+    """ Opens file, reads and returns its content as a String. """
+
+    with open(filename, "r", encoding="utf-8") as file:
+        f = file.read()
+    return f
+
+
+def load_tree_root(national_rail_xml: str) -> ET.Element:
+    """ Returns root node of tree (from data.xml) """
+
+    tree = ET.ElementTree(ET.fromstring(national_rail_xml))
+    return tree.getroot()
+
+
+def reverse_tree(root: ET.Element) -> ET.Element:
+    """ Creates a new tree containing the root and its elements in reverse, to allow
+        for traversing the tree in reverse. Returns the new tree by its root. """
+
+    reversed_elements = list(reversed(list(root.iter())))
+    new_root = ET.Element("root")
+
+    for element in reversed_elements:
+        new_root.append(element)
+
+    return new_root
+
+
+def get_incidents(root: str, namespaces: dict) -> ET.Element:
+    """ Extracts each incident from the NationalRail data and returns it. """
+
+    return root.findall('ns:PtIncident', namespaces)
+
+
+def find_text_element(element: ET.Element, path: str, namespaces: str) -> str:
+    """ Searches for element by path and namespace, returns text if element exists. """
+
+    element_found = element.find(path, namespaces)
+    return element_found.text if element_found is not None else None
+
+
+def find_all_text_elements(element: ET.Element, path: str, namespaces: str) -> list[str]:
+    """ Searches for all elements by path and namespace, returns list of elements
+        as text if they exist. """
+
+    elements_found = element.findall(path, namespaces)
+    return [e.text for e in elements_found if e.text]
+
+
+def convert_html_to_text(html_text: str) -> str:
+    """ Extracts String from html text and returns it."""
+
+    soup = BeautifulSoup(html_text, 'lxml')
+    return soup.get_text(separator=' ', strip=True)
+
+
+def convert_to_datetime(time: str) -> datetime:
+    """ Converts string in datetime format into a datetime object and returns it. """
+
+    return datetime.fromisoformat(time.replace('Z', ''))
```

Comment:
Do you not want the `Z` at the end?

Reply:
No because when comparing with .now(), I need to include timezone but datetime.now() is already in the correct time zone. Otherwise I have to use timezone UTC which gives issues when comparing with BST and GMT.